exchanges/coinbasepro/coinbasepro_websocket.go

```diff
@@ -400,96 +333,174 @@ func (c *CoinbasePro) GenerateDefaultSubscriptions() ([]subscription.Subscriptio
 	}
-	err = c.Websocket.Conn.SendJSONMessage(subscribe)
+	creds, err := c.GetCredentials(ctx)
 	if err != nil {
-		return err
+		return "", err
 	}
-	c.Websocket.AddSuccessfulSubscriptions(channelsToSubscribe...)
-	return nil
+	block, _ := pem.Decode([]byte(creds.Secret))
+	if block == nil {
+		return "", errCantDecodePrivKey
+	}
+	key, err := x509.ParseECPrivateKey(block.Bytes)
+	if err != nil {
+		return "", err
+	}
+	nonce, err := common.GenerateRandomString(64, "1234567890ABCDEF")
+	if err != nil {
+		return "", err
+	}
+	head := map[string]interface{}{"kid": creds.ClientID, "typ": "JWT", "alg": "ES256", "nonce": nonce}
+	headJSON, err := json.Marshal(head)
+	if err != nil {
+		return "", err
+	}
+	headEncode := base64URLEncode(headJSON)
+	c.jwtLastRegen = time.Now()
+	body := map[string]interface{}{"iss": "coinbase-cloud", "nbf": time.Now().Unix(),
+		"exp": time.Now().Add(time.Minute * 2).Unix(), "sub": creds.ClientID, "aud": "retail_rest_api_proxy"}
+	if uri != "" {
+		body["uri"] = uri
+	}
+	bodyJSON, err := json.Marshal(body)
+	if err != nil {
+		return "", err
+	}
+	bodyEncode := base64URLEncode(bodyJSON)
+	hash := sha256.Sum256([]byte(headEncode + "." + bodyEncode))
+	sig, err := ecdsa.SignASN1(rand.Reader, key, hash[:])
+	if err != nil {
+		return "", err
+	}
+	sigEncode := base64URLEncode(sig)
+	return headEncode + "." + bodyEncode + "." + sigEncode, nil
 }
 
-// Unsubscribe sends a websocket message to stop receiving data from the channel
-func (c *CoinbasePro) Unsubscribe(channelsToUnsubscribe []subscription.Subscription) error {
-	unsubscribe := WebsocketSubscribe{
-		Type: "unsubscribe",
+// getTimestamp is a helper function which pulls a RFC3339-formatted timestamp from a byte slice
+// of JSON data
+func getTimestamp(rawData []byte) (time.Time, error) {
+	data, _, _, err := jsonparser.Get(rawData, "timestamp")
+	if err != nil {
+		return time.Time{}, err
 	}
-	productIDs := make([]string, 0, len(channelsToUnsubscribe))
-	for i := range channelsToUnsubscribe {
-		p := channelsToUnsubscribe[i].Pair.String()
-		if p != "" && !common.StringDataCompare(productIDs, p) {
-			// get all unique productIDs in advance as we generate by channels
-			productIDs = append(productIDs, p)
-		}
+	timestamp, err := time.Parse(time.RFC3339, string(data))
+	if err != nil {
+		return time.Time{}, err
 	}
+	return timestamp, nil
+}
 
-unsubscriptions:
-	for i := range channelsToUnsubscribe {
-		for j := range unsubscribe.Channels {
-			if unsubscribe.Channels[j].Name == channelsToUnsubscribe[i].Channel {
-				continue unsubscriptions
-			}
-		}
-
-		unsubscribe.Channels = append(unsubscribe.Channels, WsChannels{
-			Name:       channelsToUnsubscribe[i].Channel,
-			ProductIDs: productIDs,
-		})
+// sendRequest is a helper function which sends a websocket message to the Coinbase server
+func (c *CoinbasePro) sendRequest(msgType, channel string, productIDs currency.Pairs) error {
+	creds, err := c.GetCredentials(context.Background())
```

Comment:
I believe this is causing issues. I do not have credentials set in my config and `authenticatedWebsocketApiSupport` is set to `false`. So I am receiving the following error when attempting to subscribe:
`[ERROR]	 | WEBSOCKET | 26/03/2024 15:20:05 | CoinbasePro websocket: subscription failure, CoinbasePro REST or Websocket authentication support is not enabled`

I do not believe authentication is required to subscribe to a ticker

Reply:
It is. All websocket endpoints require authentication. That's not explicitly stated, but these three links imply it https://docs.cloud.coinbase.com/advanced-trade-api/docs/auth https://docs.cloud.coinbase.com/advanced-trade-api/docs/ws-auth https://docs.cloud.coinbase.com/advanced-trade-api/docs/ws-channels and removing the signature/key from the request causes the exchange to complain, even for channels like "ticker".